std/experimental/color/rgb.d

```diff
@@ -0,0 +1,852 @@
+module std.experimental.color.rgb;
+
+import std.experimental.color;
+import std.experimental.color.xyz;
+
+import std.traits: isInstanceOf, isNumeric, isIntegral, isFloatingPoint, isSigned, isSomeChar, Unqual;
+import std.typetuple: TypeTuple;
+import std.typecons: tuple;
+
+@safe: pure: nothrow: @nogc:
+
+
+enum isRGB(T) = isInstanceOf!(RGB, T);
+
+enum RGBColorSpace
+{
+    sRGB,
+    sRGB_Gamma2_2,
+/+ TODO: these all need testing
+    AdobeRGB,
+    AppleRGB,
+    BestRGB,
+    BetaRGB,
+    BruceRGB,
+    CIE_RGB,
+    ColorMatchRGB,
+    DonRGB4,
+//    ECI_RGB_v2,
+    EktaSpacePS5,
+    NTSC_RGB,
+    PAL_SECAM_RGB,
+    ProPhotoRGB,
+    SMPTE_C_RGB,
+    WideGamutRGB
++/
+}
+
+
+struct RGB(string components_, ComponentType_, bool linear_ = false, RGBColorSpace colorSpace_ = RGBColorSpace.sRGB) if(isValidComponentType!ComponentType_)
+{
+@safe: pure: nothrow: @nogc:
+
+    // RGB colors may only contain components 'rgb', or 'l' (luminance)
+    // They may also optionally contain an 'a' (alpha) component, and 'x' (unused) components
+    static assert(AllIn!("rgblax", components), "Invalid Color component '"d ~ NotIn!("rgblax", components) ~ "'. RGB colors may only contain components: r, g, b, l, a, x"d);
+    static assert(AnyIn!("rgbal", components), "RGB colors must contain at least one component of r, g, b, l, a.");
+    static assert(!CanFind!(components, 'l') || !AnyIn!("rgb", components), "RGB colors may not contain rgb AND luminance components together.");
+
+    // create members for some useful information
+    alias ComponentType = ComponentType_;
+    enum string components = components_;
+    enum RGBColorSpace colorSpace = colorSpace_;
+    enum bool linear = linear_;
+    enum bool hasAlpha = CanFind!(components, 'a');
+
+
+    // mixin will emit members for components
+    template Components(string components)
+    {
+        static if(components.length == 0)
+            enum Components = "";
+        else
+            enum Components = (components[0] == 'x' ? "private " : "") ~ ComponentType.stringof ~ ' ' ~ components[0] ~ " = 0;\n" ~ Components!(components[1..$]);
+    }
+    mixin(Components!components);
+
+    // missing components will be fabricated such that code that references them compiles
+    mixin template MissingComponent(char name)
+    {
+        // missing color components will equal the luminance, if it exists. 0 otherwise.
+        static if(!CanFind!(components, 'l'))
+            mixin("enum ComponentType "~name~" = 0;");
+        else
+            mixin("alias "~name~" = l;");
+    }
+    static if(!CanFind!(components, 'r'))
+        mixin MissingComponent!'r';
+    static if(!CanFind!(components, 'g'))
+        mixin MissingComponent!'g';
+    static if(!CanFind!(components, 'b'))
+        mixin MissingComponent!'b';
+    static if(!CanFind!(components, 'l'))
+        @property ComponentType l() const { return toGrayscale!colorSpace(r, g, b); }
```

Comment:
It's complex because now there's no guarantees made.
I just kinda liked the simplicity of a reliable prescription that RGB colours would all have a reliable set of members...